File: 0102-binary-tree-level-order-traversal/0102-binary-tree-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:

        ans = []

        def helper(node, depth):
            if node:
                if len(ans) == depth:
                    ans.append([])
                ans[depth].append(node.val)
                helper(node.left, depth+1)
                helper(node.right, depth+1)
        helper(root,0)
        return ans


    def levelOrder2(self, root: Optional[TreeNode]) -> List[List[int]]:
        
        queue = deque([(root, 0)])
        ans = []

        while queue:
            node, depth = queue.popleft()
            if node:
                if len(ans) == depth:
                    ans.append([])

                ans[depth].append(node.val)

                if node.left:
                    queue.append((node.left, depth+1))
                if node.right:
                    queue.append((node.right, depth+1))

        return ans

'''
    inorder - left root right
    preorder - root left right
    postorder - left right root
    levelorder - root left right
'''

File: 0102-binary-tree-level-order-traversal/0102-binary-tree-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> levels;
        std::queue<TreeNode*> queue;
        if(!root) {
            return {};
        }

        queue.push(root);
        while(!queue.empty()) {
            vector<int> curr;
            int n = queue.size();
            for(int i=0; i<n; i++) {
                TreeNode *node = queue.front();
                curr.push_back(node->val);
                queue.pop();
                if(node->left) queue.push(node->left);
                if(node->right) queue.push(node->right);
            }
            levels.push_back(curr);
        }

        return levels;
    }
};

/*
stack FIFO
queue LIFO
add node
add left right

*/